Rationale.md

```diff
@@ -156,22 +156,29 @@ TODO
 
 ## Indirect Calls
 
-The exact semantics of indirect function calls, function pointers, and what
-happens when calling the wrong function, are still being discussed.
-
-Fundamentally linear memory is a simple collection of bytes, which means that
-some integral representation of function pointers must exist. It's desirable to
-hide the actual address of generated code from untrusted code because that would
-be an unfortunate information leak which could have negative security
-implications. Indirection is therefore desired.
-
-One extra concern is that existing C++ code sometimes stores data inside of what
-is usually a function pointer. This is expected to keep working.
-
-Dynamic linking further complicates this: WebAssembly cannot simply standardize
-on fixed-size function tables since dynamically linked code can add new
-functions, as well as remove them.
-
+The table-based scheme for indirect function calls was motivated by the need
+to represent function pointers as integer values that can be stored into the
+linear memory, as well as to enforce basic safety properties such as a
+calling a function with the wrong signature does not destroy the safety
+guarantees of WebAssembly.
+
+Languages like C and C++ that compile to WebAssembly also imposed
+requirements, such as the uniqueness of function pointers and the ability
+to compare function pointers to data pointers, or treat data as function
+pointers.
+
+Several alternatives to direct indices with a heterogeneous indirect function table
+were considered, from alternatives with multiple tables to statically typed function
+pointers that can be mapped back and forth to integers. With the added complication
+of dynamic linking, none of these alternatives perfectly fit the requirements.
+
+The current design requires two dynamic checks when invoking a function pointer:
+a bounds check against the size of the indirect function table and a signature check
+for the function at that index against an expected signature. Some dynamic optimization
+techniques (e.g. inline caches, or a one-element cache), can reduce the number of
```

Comment:
I'd also mention that engines can internally use table-per-signature internally as an implementation strategy.